repotracker/repotracker.go

```diff
@@ -693,130 +705,127 @@ func sanityCheckOrderNum(revOrderNum int, projectId, revision string) error {
+
+	return client.UseSession(ctx, func(sessCtx mongo.SessionContext) error {
+		// generate all task Ids so that we can easily reference them for dependencies
+		sourceRev := ""
+		if metadata.SourceVersion != nil {
+			sourceRev = metadata.SourceVersion.Revision
 		}
-	}
-	taskIds := model.NewTaskIdTable(project, v, sourceRev, metadata.DefinitionID)
+		taskIds := model.NewTaskIdTable(project, v, sourceRev, metadata.DefinitionID)
 
-	// create all builds for the version
-	for _, buildvariant := range project.BuildVariants {
-		if buildvariant.Disabled {
-			continue
+		err := sessCtx.StartTransaction()
+		if err != nil {
+			return errors.Wrap(err, "error starting transaction")
 		}
-		var match bool
-		if len(aliases) > 0 {
-			match, err = aliases.HasMatchingVariant(buildvariant.Name)
-			if err != nil {
-				grip.Error(err)
+		// create all builds for the version
+		for _, buildvariant := range project.BuildVariants {
+			if buildvariant.Disabled {
 				continue
 			}
-			if !match {
-				continue
+			var match bool
+			if len(aliases) > 0 {
+				match, err = aliases.HasMatchingVariant(buildvariant.Name)
+				if err != nil {
+					grip.Error(err)
+					continue
+				}
+				if !match {
+					continue
+				}
 			}
-		}
-		args := model.BuildCreateArgs{
-			Project:      *project,
-			Version:      *v,
-			TaskIDs:      taskIds,
-			BuildName:    buildvariant.Name,
-			Activated:    false,
-			SourceRev:    sourceRev,
-			DefinitionID: metadata.DefinitionID,
-			Aliases:      aliases,
-		}
-		var buildId string
-		buildId, err = model.CreateBuildFromVersion(args)
-		if err != nil {
-			if metadata.TriggerID == "" || !strings.Contains(err.Error(), "E11000") {
-				grip.Error(message.WrapError(err, message.Fields{
-					"message": "error inserting build",
-					"version": v.Id,
-					"variant": buildvariant.DisplayName,
-				}))
+			args := model.BuildCreateArgs{
+				Project:      *project,
+				Version:      *v,
+				TaskIDs:      taskIds,
+				BuildName:    buildvariant.Name,
+				Activated:    false,
+				SourceRev:    sourceRev,
+				DefinitionID: metadata.DefinitionID,
+				Aliases:      aliases,
+				Session:      sessCtx,
+			}
+			var buildId string
+			buildId, err = model.CreateBuildFromVersion(args)
+			if err != nil {
+				abortErr := sessCtx.AbortTransaction(sessCtx)
+				grip.Notice(message.Fields{
+					"message":    "aborting transaction",
+					"cause":      "can't insert build items",
+					"variant":    buildvariant.Name,
+					"version":    v.Id,
+					"insert_err": err.Error(),
+					"abort_err":  abortErr.Error(),
+				})
+				return errors.Wrapf(err, "error inserting build %s", buildId)
 			}
-			continue
-		}
 
-		var lastActivated *model.Version
-		lastActivated, err = model.VersionFindOne(model.VersionByLastVariantActivation(ref.Identifier, buildvariant.Name))
-		if err != nil {
-			return errors.Wrap(err, "problem getting activatation time for variant")
-		}
+			var lastActivated *model.Version
```

Comment:
Should the below return abort the transaction? (`lastActivated, err = model.VersionFindOne(model.VersionByLastVariantActivation(ref.Identifier, buildvariant.Name))
`)

Reply:
Doh, added a continue on error here